Make resampling symmetric

fepegar opened the pull request:
<!-- Replace {issue_number} with the issue that will be closed after merging this PR.
For example: Fixes #37.
If there isn't one, delete the line below. -->

**Description**

<!-- Write a few sentences describing the changes proposed in this pull request. -->

INTERACTIVE DEMO: https://editor.p5js.org/fepegar/full/bsD299_fU

The current implementation aligns the input and output images at the boundaries closest to the origin index (here, bottom right, downsampling from 1 mm to 2.3 mm, green is input and magenta is output):

![image](https://github.com/user-attachments/assets/205b9538-7032-48e2-bac7-8a62fb322627)

This causes that e.g. `flip(resample(x)) != resample(flip(x))` (the new unit test fails on `main` and not here).

In this PR, I implemented something closer to grid 2 in https://ppwwyyxx.com/blog/2021/Where-are-Pixels/, where input and output are aligned at their geometrical center. Resampling the image above looks like this:

![image](https://github.com/user-attachments/assets/2ed728b2-528f-49b0-a9a3-89cfeabe201f)

**Checklist**

<!-- You do not need to complete all the items by the time you submit the pull
request, but most likely the changes will only be merged if all the tasks are
done. See more information about the submission process in the
CONTRIBUTING (https://github.com/TorchIO-project/torchio/blob/main/CONTRIBUTING.rst) docs. -->

<!-- Write an `x` in all the boxes that apply -->
- [x] I have read the [`CONTRIBUTING`](https://github.com/TorchIO-project/torchio/blob/main/CONTRIBUTING.rst) docs and have a developer setup ready
- Changes are
  - [x] Non-breaking (would not break existing functionality)
  - [ ] Breaking (would cause existing functionality to change)
- [x] Tests added or modified to cover the changes
- [ ] In-line docstrings updated
- [ ] Documentation updated
- [x] This pull request is ready to be reviewed

romainVala commented:
Hi Fernando
I just see this PR, but I am not sure to understand ...
in the link you gave they said that  scikit-image transform.resize perform the 2 strategie
So this is what we should have here ...

romainVala commented:
not sure if it is related, but I encounter issue with half voxel shift here
https://github.com/nipy/nibabel/issues/1366

romainVala commented:
> **Description**
> 
> INTERACTIVE DEMO: https://editor.p5js.org/fepegar/full/bsD299_fU

Nice demo !

fepegar commented:
> in the link you gave they said that scikit-image transform.resize perform the 2 strategie
So this is what we should have here ...

That's what I implemented in this PR! From the description:

> In this PR, I implemented something closer to grid 2 in [ppwwyyxx.com/blog/2021/Where-are-Pixels](https://ppwwyyxx.com/blog/2021/Where-are-Pixels/)

fepegar commented:
I'll merge now as we seem to agree that grid 2 is the desirable one.
